src/misc/param.cc

```diff
@@ -61,21 +61,29 @@ void initEnv() {
 
 void ncclLoadParam(char const* env, int64_t deftVal, int64_t uninitialized, int64_t* cache) {
   static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+  if (__builtin_expect(__atomic_load_n(cache, __ATOMIC_RELAXED) != uninitialized, true)) {
+    return;
+  }
+
```

Comment:
This shouldn't be necessary. You are basically duplicating what's already done by the caller of this function, which is: https://github.com/NVIDIA/nccl/blob/ab2b89c4c339bd7f816fbc114a4b05d386b66290/src/include/param.h#L24-L26

Reply:
> This shouldn't be necessary. You are basically duplicating what's already done by the caller of this function, which is:
> 
> https://github.com/NVIDIA/nccl/blob/ab2b89c4c339bd7f816fbc114a4b05d386b66290/src/include/param.h#L24-L26

It seems so. However, as an atomic API, wouldn't it be better to have this logic inside the API?

Reply:
Maintainability-wise, I agree with you that it would make sense to hide it inside the function. However, NCCL has these calls sprinkled all over the codebase, including in performance-critical places. Looking at the code history, I can see that the code was deliberately structured this way to minimize the runtime impact.

Reply:
> Maintainability-wise, I agree with you that it would make sense to hide it inside the function. However, NCCL has these calls sprinkled all over the codebase, including in performance-critical places. Looking at the code history, I can see that the code was deliberately structured this way to minimize the runtime impact.

Yes, I see what you mean, and addressed your comment.